bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/link/Link.java

```diff
@@ -0,0 +1,76 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2019 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.link;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jetbrains.annotations.Nullable;
+
+import com.adobe.cq.wcm.core.components.models.mixin.LinkMixin;
+import com.day.cq.wcm.api.Page;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Wraps link information to be used in models.
+ */
+public final class Link implements LinkMixin {
+
+    private final String linkURL;
+    private final Map<String, String> linkHtmlAttributes;
+    private final Page targetPage;
+
+    Link(String linkURL, String linkTarget, Page targetPage) {
+        this.linkURL = linkURL;
+        this.linkHtmlAttributes = buildAttributes(linkURL, linkTarget);
+        this.targetPage = targetPage;
+    }
+
+    @Override
+    public boolean isLinkValid() {
+        return linkURL != null;
+    }
+
+    @Override
+    public @Nullable String getLinkURL() {
+        return linkURL;
+    }
+
+    @Override
+    public @Nullable Map<String, String> getLinkHtmlAttributes() {
+        return linkHtmlAttributes;
+    }
+
+    /**
+     * @return Target page if the link URL pointed to an internal page.
+     */
+    public Page getTargetPage() {
```

Comment:
Nullable?

Reply:
+1, forgot it. i'll wait if other review comments are posted and will update it then.

Reply:
Why not use Optional<Page> to express this in the code as well?

Reply:
thought about this as well and even started to use it for this method but refactored back as it's still unusual to be used in the sling/AEM/core component APIs - and it will not work well in HTL (although this Link class is currently not used in HTL at all).
so i'm completly open to use Optional in this case, but it would be the first place in all the existing APIs, and i thought that was not worth it.

Reply:
HTL support is a good topic, indeed.

But being the first one ... At some point we have to use it. Because I see way too much code with NPEs and people not using linting tools and also not reading JavaDoc. And using Optional forces them to reason about it and makes it really implicit that nothing can be returned (not `null`).